turbinia/lib/libcloudforensics.py

```diff
@@ -437,64 +437,126 @@ def get_or_create_analysis_vm(
+      which returns a dictionary with name and metadata(zone, labels)
+      for each disk.
+      ex: {'disk-1': {'zone': 'us-central1-a', 'labels': {'id': '12
+    """
+
+    disk_service_object = self.gce_api().disks()
+    return self.__list_by_label(labels_filter, disk_service_object, 'OR')
+
+  def __list_by_label(self, labels_filter, service_object, operation):
+    """lists Disks/VMs in a project with one/all of the provided labels.
+
+    The operation parameter will determine if the labels will be ored or anded.
+
+    Args:
+      labels_filter:  a dict of labels to find --> {'id': '123'}
       service_object: Google Compute Engine (Disk | Instance) service object
+      operation: a string with the value "OR" or "AND" determining how the
+      labels should be compined.
 
     Returns:
-      Dictionary with name and metadata(zone, labels) for each instance.
+      Dictionary with name and metadata(zone, labels) for each instance/disk.
       ex: {'instance-1': {'zone': 'us-central1-a', 'labels': {'id': '123'}}
     """
+    if operation not in ['OR', 'AND']:
+      raise RuntimeError((
+          '__list_by_label third argument must be "OR" or "AND"'
+          ' {0:s} is an invalid argument.').format(operation))
+
     resource_dict = dict()
     filter_expression = ''
     for key, value in labels_filter.items():
-      filter_expression = 'labels.' + key + '=' + value
-      request = service_object.aggregatedList(
-          project=self.project_id, filter=filter_expression)
-      while request is not None:
-        response = request.execute()
-        result = self.gce_operation(response, zone=self.default_zone)
-
-        for region_or_zone_string, resource_scoped_list in result[
-            'items'].items():
-          for resource_key in resource_scoped_list:
-            if resource_key in ['disks', 'instances']:
-              for item in resource_scoped_list[resource_key]:
-                _, zone = region_or_zone_string.rsplit('/', 1)
-                resource_dict[item['name']] = dict(
-                    zone=zone, labels=item['labels'])
-
-        request = service_object.aggregatedList_next(
-            previous_request=request, previous_response=response)
+      filter_expression += 'labels.{0:s}={1:s} {2:s} '.format(
+          key, value, operation)
+    filter_expression = filter_expression[:-(len(operation) + 1)]
+
+    request = service_object.aggregatedList(
+        project=self.project_id, filter=filter_expression)
+    while request is not None:
+      response = request.execute()
+      result = self.gce_operation(response, zone=self.default_zone)
+
+      for region_or_zone_string, resource_scoped_list in result['items'].items(
+      ):
+
+        if 'warning' not in resource_scoped_list.keys():
+          _, zone = region_or_zone_string.rsplit('/', 1)
+          for item in resource_scoped_list.get(
+              'instances', []) or resource_scoped_list.get('disks', []):
```

Comment:
the aggregatedList request is invoked either using a gce_api().disks() or gce_api().instances() service object (the service_object parameter passed by the calling method) that's why i works as intended since it will always be either instances only or disks only. 
I thought about multiple statements, however even if we invoke the aggregatedList request on instances and disks each time, we will need to have two returned dicts since any disk with the same name and zone as the instance (default for boot disks) will overwrite the instance in the resource_dict. But (instance | disk) names are unique to other (instance | disk) names in the same project. Does that make sense for you?

Reply:
Ah yeah, that makes sense, thanks for clarifying! Even if the way the API works means this is working as intended, it looks a bit weird from a code perspective. I'd still suggest to split this into two lines similar to the above, and adding a comment specifying that only `'instances'` or `'disks'` will be in that dict. It will have the same effect but IMO is easier to read and understand what's going on. What do you think?

Reply:
yes, that make sense, just pushed the new change.